language-plutus-core/src/Language/PlutusCore/TypeRenamer.hs

```diff
@@ -1,120 +1,182 @@
+annotateType (TyVar x (TyName (Name x' b (Unique u)))) = do
+    st <- gets _types
+    case IM.lookup u st of
+        Just ty -> pure $ TyVar x (TyNameWithKind (TyName (Name (x', ty) b (Unique u))))
+        Nothing -> throwError $ UnboundVar (Name x' b (Unique u))
+annotateType (TyLam x (TyName (Name x' s u@(Unique i))) k ty) = do
+    insertKind i k
+    let nwty = TyNameWithKind (TyName (Name (x', k) s u))
+    TyLam x nwty k <$> annotateType ty
+annotateType (TyForall x (TyName (Name x' s u@(Unique i))) k ty) = do
+    insertKind i k
+    let nwty = TyNameWithKind (TyName (Name (x', k) s u))
+    TyForall x nwty k <$> annotateType ty
+annotateType (TyFix x (TyName (Name x' s u@(Unique i))) k ty) = do
+    insertKind i k
+    let nwty = TyNameWithKind (TyName (Name (x', k) s u))
+    TyFix x nwty k <$> annotateType ty
+annotateType (TyFun x ty ty') =
+    TyFun x <$> annotateType ty <*> annotateType ty'
+annotateType (TyApp x ty tys) =
+    TyApp x <$> annotateType ty <*> traverse annotateType tys
+annotateType (TyBuiltin x tyb) = pure (TyBuiltin x tyb)
 
 -- This renames terms so that they have a unique identifier. This is useful
 -- because of scoping.
-rename :: IdentifierState -> Term a -> Term a
-rename st = flip evalState st . renameTerm
-
--- TODO use an anamorphism?
-renameTerm :: Term a -> IdentifierM (Term a)
-renameTerm t@(LamAbs x (Name x' s (Unique u)) t') = do
-    pastDef <- gets (IM.lookup u . fst)
-    m <- gets (fst . IM.findMax . fst)
+rename :: IdentifierState -> Program TyName Name a -> Program TyName Name a
+rename (st, _) (Program x v p) = Program x v (evalState (renameTerm (Identifiers st') p) m)
+    where st' = IM.fromList (zip keys keys)
+          keys = IM.keys st
+          m = fst (IM.findMax st)
+
+newtype Identifiers = Identifiers { _identifiers :: IM.IntMap Int }
+    deriving Show
+
+type MaxM = State Int
+
+identifiers :: Lens' Identifiers (IM.IntMap Int)
+identifiers f s = fmap (\x -> s { _identifiers = x }) (f (_identifiers s))
+
+modifyIdentifiers :: Int -> Int -> Identifiers -> Identifiers
+modifyIdentifiers u m = over identifiers (IM.insert u (m+1) . IM.insert (m+1) (m+1))
+
+lookupId :: Int -> Identifiers -> Maybe Int
+lookupId u st = IM.lookup u (_identifiers st)
+
+-- this convoluted affair lets us track the maximum in a global state monad,
+-- while keeping the table for renaming local (so that we don't rename things in
+-- function applications)
+renameTerm :: Identifiers -> Term TyName Name a -> MaxM (Term TyName Name a)
+renameTerm st t@(LamAbs x (Name x' s (Unique u)) ty t') = do
+    m <- get
+    let st' = modifyIdentifiers u m st
+        pastDef = lookupId u st
     case pastDef of
-        Just _ -> LamAbs x (Name x' s (Unique $ m+1)) <$> rewriteWith (Unique u) (Unique $ m+1) t'
+        Just _ ->
+            modify (+1) >>
+            LamAbs x (Name x' s (Unique (m+1))) <$> renameType st' ty <*> renameTerm st' t'
         _      -> pure t
```

Comment:
I don't think that is correct. Even if `pastDef == Nothing`, we still need to descend into `t'` as there may be other binders to rename in that subtree, don't we? The same holds for `ty`. So, the difference between the `Just` and the `Nothing` branch is really only whether we change the unique (and perform `modify (+1)`). Have I overlooked anything?